src/keys_handler.rs

```diff
@@ -2,26 +2,207 @@
 // Copyright 2021 Keylime Authors
 
 use actix_web::{web, HttpResponse, Responder};
+use log::*;
 use serde::Deserialize;
 
+use json::parse;
+use openssl::{
+    aes::AesKey,
+    encrypt::{Decrypter, Encrypter},
+    pkey::{PKey, Private, Public},
+    rsa::{Padding, Rsa},
+    symm::{decrypt_aead, Cipher},
+};
+
+use crate::{error::Error, QuoteData};
+
+// This value comes from and must match the Python codebase
+// https://github.com/keylime/keylime/blob/1ed43ac8f75d5c3bc3a3bbbbb5037f20cf3c5a6a/ \
+// keylime/crypto.py#L23
+const AES_BLOCK_SIZE: usize = 16;
+
+// TODO: perhaps there is a better way to do globals
+static mut UKEY: [u8; 32] = [0u8; 32];
+static mut VKEY: [u8; 32] = [0u8; 32];
+static mut PAYLOAD_SYM_KEY: [u8; 32] = [0u8; 32];
+
 #[derive(Deserialize)]
 pub struct Verify {
     challenge: String,
 }
 
-#[derive(Deserialize)]
-pub struct UkeyJson {
-    b64_encrypted_key: String, // this will be handled as a blog once wired in
-    auth_tag: String,
-}
-
 pub async fn verify(param: web::Query<Verify>) -> impl Responder {
     HttpResponse::Ok().body(format!("Challenge: {}", param.challenge))
 }
 
-pub async fn ukey(param: web::Json<UkeyJson>) -> impl Responder {
-    HttpResponse::Ok().body(format!(
-        "b64_encrypted_key: {} auth_tag: {}",
-        param.b64_encrypted_key, param.auth_tag
-    ))
+pub(crate) fn xor_in_place(a: &mut [u8], b: &[u8]) -> Result<(), Error> {
+    if a.len() != b.len() {
+        return Err(Error::Other("cannot xor differing length slices".to_string()));
+    }
+    for (x, y) in a.iter_mut().zip(b) {
+        *x ^= *y;
+    }
+
+    Ok(())
+}
+
+// Parameters are based on Python codebase:
+// https://github.com/keylime/keylime/blob/1ed43ac8f75d5c3bc3a3bbbbb5037f20cf3c5a6a/ \
+// keylime/crypto.py#L189
+pub fn decrypt_payload(
+    payload: &[u8],
+    symm_key: &[u8],
+) -> Result<Vec<u8>, openssl::error::ErrorStack> {
+    // parse out payload iv, tag, ciphertext
+    let (iv, rest) = payload.split_at(AES_BLOCK_SIZE);
+    let (ciphertext, tag) = rest.split_at(rest.len() - AES_BLOCK_SIZE);
+
+    // note: it's unclear in the python codebase whether this is AES 128, but 128 seems
+    // to be the default
+    // https://cryptography.io/en/latest/hazmat/primitives/symmetric-encryption/#symmetric-encryption-modes
+    decrypt_aead(
+        Cipher::aes_128_gcm(),
+        symm_key,
+        Some(iv),
+        &[0u8], // todo: check that we don't need additional auth data here?
```

Comment:
There is no additional data: that would be a call like `encryptor.authenticate_additional_data(associated_data)` in the `encrypt` function, or `decryptor.authenticate_additional_data(authenticated_data)` in `decrypt`.